File: backend/internal/handler/admin_pool.go
package handler

import (
	"context"
	"encoding/json"
	"log"
	"net/http"

	"github.com/eriksteenman/reign-game/backend/internal/httperr"
	poolsvc "github.com/eriksteenman/reign-game/backend/internal/service/pool"
)

// PoolService is the application surface the admin pool handler depends
// on. Orchestration (fetching configs, counting ready puzzles, emitting
// per-step DDB timing) lives in internal/service/pool.
type PoolService interface {
	LoadPool(ctx context.Context) ([]poolsvc.ComboEntry, error)
}

// ComboStatus is the per-combo entry in the pool status response.
// Config nests a ConfigBody (shared with the flat ConfigView DTO) so
// there is one canonical shape for the threshold/enabled/maxAttempts
// fields across every CONFIG-bearing API surface.
type ComboStatus struct {
	Size       int        `json:"size"`
	Mode       string     `json:"mode"`
	Config     ConfigBody `json:"config"`
	ReadyCount int        `json:"readyCount"`
}

type adminPoolResponse struct {
	Combos []ComboStatus `json:"combos"`
}

// AdminPoolHandler creates an HTTP handler for GET /admin/pool.
// It delegates orchestration to the pool service and maps the result
// to a JSON response.
func AdminPoolHandler(svc PoolService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		entries, err := svc.LoadPool(r.Context())
		if err != nil {
			log.Printf("admin pool handler: LoadPool failed: %v", err)
			httperr.WriteError(w, http.StatusInternalServerError, "internal_error", "Failed to retrieve pool status.")
			return
		}

		combos := make([]ComboStatus, 0, len(entries))
		for _, e := range entries {
			combos = append(combos, ComboStatus{
				Size: e.Config.Size,
				Mode: e.Config.Mode,
				Config: ConfigBody{
					Threshold:   e.Config.Threshold,
					Enabled:     e.Config.Enabled,
					MaxAttempts: e.Config.MaxAttempts,
				},
				ReadyCount: e.ReadyCount,
			})
		}

		resp := adminPoolResponse{Combos: combos}
		if err := json.NewEncoder(w).Encode(resp); err != nil {
			log.Printf("admin pool handler: failed to write response: %v", err)
		}
	}
}

File: frontend/playwright/test-helpers/admin-pool.ts
import { expect, type APIRequestContext } from "@playwright/test";

/**
 * Shared types + reader for GET /api/admin/pool. Backend source of truth
 * is `backend/internal/handler/config_dto.go::ConfigBody` and the admin
 * pool handler's response shape. `maxAttempts` is `json:"...,omitempty"`
 * on the backend so it is optional here.
 */

export interface ConfigBody {
  threshold: number;
  enabled: boolean;
  maxAttempts?: number;
}

export interface ComboStatus {
  size: number;
  mode: string;
  config: ConfigBody;
  readyCount: number;
}

export interface AdminPoolResponse {
  combos: ComboStatus[];
}

/**
 * Read the `size#mode` combo from GET /api/admin/pool. Throws loudly if
 * the combo is missing so a fixture/seed drift fails the test instead
 * of silently treating the absence as `readyCount = 0`.
 */
export async function readCombo(
  request: APIRequestContext,
  size: number,
  mode: string,
): Promise<ComboStatus> {
  const res = await request.get("/api/admin/pool");
  expect(res.ok(), `GET /api/admin/pool failed: ${res.status()}`).toBe(true);
  const body = (await res.json()) as AdminPoolResponse;
  const combo = body.combos.find((c) => c.size === size && c.mode === mode);
  if (!combo) {
    throw new Error(
      `combo ${size}#${mode} missing from /api/admin/pool response — ` +
        `seed/CONFIG fixture likely not loaded`,
    );
  }
  return combo;
}
